dali/operators/image/convolution/laplacian.cc

```diff
@@ -0,0 +1,222 @@
+
+  bool SetupImpl(std::vector<OutputDesc>& output_desc, const workspace_t<CPUBackend>& ws) override {
+    const auto& input = ws.template Input<CPUBackend>(0);
+    int nsamples = input.num_samples();
+    auto nthreads = ws.GetThreadPool().NumThreads();
+
+    output_desc.resize(1);
+    output_desc[0].type = type2id<Out>::value;
+    output_desc[0].shape.resize(nsamples, input.shape().sample_dim());
+
+    args.ObtainLaplacianArgs(ws, nsamples);
+    windows_.resize(nsamples);
+
+    for (int sample_idx = 0; sample_idx < nsamples; sample_idx++) {
+      const auto& window_sizes = args.GetWindowSizes(sample_idx);
+      for (int i = 0; i < axes; i++) {
+        for (int j = 0; j < axes; j++) {
+          windows_[sample_idx][i][j] = lap_windows_.GetWindow(window_sizes[i][j], i == j);
+        }
+      }
+    }
+
+    kmgr_.template Resize<Kernel>(nthreads, nsamples);
+    for (int sample_idx = 0; sample_idx < nsamples; sample_idx++) {
+      // We take only last `ndim` siginificant dimensions to handle sequences as well
+      auto elem_shape = input[sample_idx].shape().template last<ndim>();
+      kmgr_.Setup<Kernel>(sample_idx, ctx_, elem_shape, args.GetWindowSizes(sample_idx));
+      // The shape of data stays untouched
+      output_desc[0].shape.set_tensor_shape(sample_idx, input[sample_idx].shape());
+    }
+    return true;
+  }
+
+  void RunImpl(workspace_t<CPUBackend>& ws) override {
+    const auto& input = ws.template Input<CPUBackend>(0);
+    auto& output = ws.template Output<CPUBackend>(0);
+    output.SetLayout(input.GetLayout());
+    auto in_shape = input.shape();
+    auto& thread_pool = ws.GetThreadPool();
+
+    int nsamples = input.shape().num_samples();
+    for (int sample_idx = 0; sample_idx < nsamples; sample_idx++) {
+      const auto& shape = input[sample_idx].shape();
+      auto elem_volume = volume(shape.begin() + dim_desc_.usable_axes_start, shape.end());
+      auto priority = elem_volume * args.GetTotalWindowSizes(sample_idx);
+
+      int seq_elements = 1;
+      int64_t stride = 0;
+      if (dim_desc_.is_sequence) {
+        seq_elements = volume(shape.begin(), shape.begin() + dim_desc_.usable_axes_start);
+        stride = elem_volume;
+      }
+      for (int elem_idx = 0; elem_idx < seq_elements; elem_idx++) {
+        thread_pool.AddWork(
+            [this, &input, &output, sample_idx, elem_idx, stride](int thread_id) {
+              const auto& scales = args.GetScales(sample_idx);
+              auto elem_shape = input[sample_idx].shape().template last<ndim>();
+              auto in_view = TensorView<StorageCPU, const In, ndim>{
+                  input[sample_idx].template data<In>() + stride * elem_idx, elem_shape};
+              auto out_view = TensorView<StorageCPU, Out, ndim>{
+                  output[sample_idx].template mutable_data<Out>() + stride * elem_idx, elem_shape};
+              // Copy context so that the kernel instance can modify scratchpad
+              auto ctx = ctx_;
```

Comment:
Shouldn't we have a `nthreads` number of ctxes?
How kernel can modify the scrachpad?

Reply:
I copied that approach from gaussian, each "piece of work" gets its own copy of the context and then kernel declares its own PreallocatedScratchpad sub_scratch that it sets to the context and passes down to sub-kernels.

Reply:
Understood.